server/request_record.go

```diff
@@ -0,0 +1,71 @@
+package server
+
+import (
+	"fmt"
+	"github.com/flashbots/rpc-endpoint/database"
+	"github.com/google/uuid"
+	"net/http"
+	"sync"
+)
+
+type requestRecord struct {
+	requestEntry        *database.RequestEntry
+	ethSendRawTxEntries []*database.EthSendRawTxEntry
+	mutex               *sync.Mutex
+	db                  database.Store
+}
+
+func NewRequestRecord(db database.Store) *requestRecord {
+	return &requestRecord{
+		requestEntry:        &database.RequestEntry{},
+		ethSendRawTxEntries: make([]*database.EthSendRawTxEntry, 0),
+		mutex:               &sync.Mutex{},
+		db:                  db,
+	}
+}
+
+func (r *requestRecord) AddEthSendRawTxEntry(id, requestId uuid.UUID) *database.EthSendRawTxEntry {
+	entry := &database.EthSendRawTxEntry{
+		Id:        id,
+		RequestId: requestId,
+	}
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	r.ethSendRawTxEntries = append(r.ethSendRawTxEntries, entry)
+	return entry
+}
+
+func (r *requestRecord) UpdateRequestEntry(req *http.Request, reqStatus int, error string) {
+	r.requestEntry.HttpMethod = req.Method
+	r.requestEntry.IpHash = GetIPHash(req)
+	r.requestEntry.Error = error
+	r.requestEntry.HttpUrl = req.URL.Path
+	r.requestEntry.HttpQueryParam = req.URL.RawQuery
+	r.requestEntry.HttpResponseStatus = reqStatus
+	r.requestEntry.Origin = req.Header.Get("Origin")
+	r.requestEntry.Host = req.Header.Get("Host")
+}
+
+// SaveRecord will insert both requestRecord and rawTxEntries to db
+func (r *requestRecord) SaveRecord() error {
+	if len(r.getForwardedRawTxEntries()) > 0 { // Save entries if the request contains rawTxEntries
+		if err := r.db.SaveRequestEntry(r.requestEntry); err != nil {
+			return fmt.Errorf("SaveRequestEntry failed %v", err)
+		}
+		if err := r.db.SaveRawTxEntries(r.ethSendRawTxEntries); err != nil {
+			return fmt.Errorf("SaveRawTxEntries failed %v", err)
+		}
+	}
+	return nil
+}
+
+// getForwardedRawTxEntries returns list of rawTxEntry which are either sent to relay or mempool or entry with error
+func (r *requestRecord) getForwardedRawTxEntries() []*database.EthSendRawTxEntry {
+	entries := make([]*database.EthSendRawTxEntry, 0, len(r.ethSendRawTxEntries))
```

Comment:
```suggestion
func (r *requestRecord) getForwardedRawTxEntries() (entries []*database.EthSendRawTxEntry) {
```

Reply:
maybe we should rename this method to `getRawTxEntriesToSave`, because it's not only forwarded ones? 🤔

Reply:
updated